content/en/blog/2022/cni-mode/index.md

```diff
@@ -0,0 +1,69 @@
+
+<!--## 为什么需要 CNI 模式？-->
+## Why CNI mode is needed
+<!--其一，之前的 Merbridge 只有一个很小的控制面，其监听 Pod 资源，将当前节点的 IP 信息写入 `local_pod_ips` 的 map，以供 connect 使用。但是，connect 程序由于工作在主机内核层，其无法知道当前正在处理的是哪个 Pod 的流量，就没法处理如 `excludeOutboundPorts` 等配置。为了能够适配注入 `excludeOutboundPorts` 的 Sidecar Annotation，我们需要让 eBPF 程序能够得知当前正在处理哪个 Pod 的请求。-->
+First, Merbridge has a small control plane, which listens to Pods and writes the current node IP into the map of `local_pod_ips` for `connect` to use. However, `connect` works in the host kernel. It needs to know a pod's identity before responding to configurations such as `excludeOutboundPorts`. In order to make Sidecar Annotations injected with `excludeOutboundPorts` effective, we need to let eBPF know the request is from which Pod.
+<!--为此，我们设计了一套方法，与 CNI 配合，能够获取当前 Pod 的 IP，以适配针对 Pod 的特殊配置。-->
+To this end, we designed a method that, in cooperation with CNI, can obtain the current Pod IP to validate special configurations for the Pod.
+<!--其二，在之前的 Merbridge 版本中，只有 connect 会处理主机发起的请求，这在同一台主机上的 Pod 互相通讯时，是没有问题的。但是在不同主机之间通讯时就会出现问题，因为按照之前的逻辑，在跨节点通讯时流量不会被修改，这会导致在接收端还是离不开 iptables。-->
+Second, in older Merbridge versions, only `connect` would process requests from the host, which was alright for same-node pod communication. However, in the case of cross-node pod communication, traffic will not be modified according to the previous logic. It means iptables are still necessary on the receiving end.
+<!--这次，我们依靠 XDP 程序，解决入口流量处理的问题。因为 XDP 程序需要挂载网卡，所以也需要借助 CNI。-->
+Here, we turned to the XDP program for ingress traffic processing. The XDP program needs to mount a network card, also making it necessary to have a CNI mode.
+<!--## CNI 如何解决问题？-->
+## How does CNI work
+<!--这里我们将探讨 CNI 的工作原理，以及如何使用 CNI 来解决问题。-->
+This section will explore how CNI works and how to use CNI to solve the issues mentioned above.
+<!--### 如何通过 CNI 让 eBPF 程序获取当前正在处理的 Pod IP？-->
+### How to use CNI to let eBPF have the current Pod IP?
+<!--我们通过 CNI，在 Pod 创建的时候，将 Pod 的 IP 信息写入一个 Map（`mark_pod_ips_map`），其 Key 为一个随机的值，Value 为 Pod 的 IP。然后，在当前 Pod 的 NetNS 里面监听一个特殊的端口 39807，将 Key 使用 `setsockopt` 写入这个端口 socket 的 mark。-->
+When a pod is created, we use CNI to write Pod IP into the map `mark_pod_ips_map` with a random value as the Key and the Pod IP as the value. Then, listen to a special port `39807` in the NetNS of the current Pod, and write the Key to the Mark of this port socket using `setsockopt`.
+<!--在 eBPF 中，我们通过 `bpf_sk_lookup_tcp` 取得端口 39807 的 Mark 信息，然后从 `mark_pod_ips_map` 中即可取得当前 NetNS（也是当期 Pod）的 IP。-->
+In eBPF, we get the Mark information of port `39807` through `bpf_sk_lookup_tcp`, and then we can get the current Pod IP, i.e., the corresponding NetNS from `mark_pod_ips_map`.
+<!--有了当前 Pod IP 之后，我们可以根据这个 Pod 的配置，确认流量处理路径（比如 `excludeOutboundPorts`）。-->
+With the current Pod IP, we can confirm the traffic processing path (such as `excludeOutboundPorts`) according to the configuration of this Pod.
```

Comment:
With the current Pod IP, we can confirm the path to route traffic (such as `excludeOutboundPorts`) according to the configuration of this Pod.